modules/images/dominant-color/load.php

```diff
@@ -0,0 +1,305 @@
+	foreach ( $filters as $filter ) {
+		add_filter( $filter, 'dominant_color_filter_content_tags', 20 );
+	}
+}
+
+/**
+ * Add CSS needed for to show the dominant color as an image background.
+ *
+ * @since n.e.x.t
+ */
+function dominant_color_add_inline_style() {
+	$handle = 'dominant-color-styles';
+	wp_register_style( $handle, false );
+	wp_enqueue_style( $handle );
+	$custom_css = 'img[data-dominantcolor]:not(.has-transparency) { background-color: var(--dominant-color); background-clip: content-box, padding-box; }';
+	wp_add_inline_style( $handle, $custom_css );
+}
+add_filter( 'wp_enqueue_scripts', 'dominant_color_add_inline_style' );
+
+
+/**
+ * Overloads wp_image_editors() to load the extended classes.
+ *
+ * @since n.e.x.t
+ *
+ * @return string[] Registered image editors class names.
+ */
+function dominant_color_set_image_editors() {
+	if ( ! class_exists( 'Dominant_Color_Image_Editor_GD' ) ) {
+		require_once 'class-dominant-color-image-editor-gd.php';
+	}
+	if ( ! class_exists( 'Dominant_Color_Image_Editor_Imagick' ) ) {
+		require_once 'class-dominant-color-image-editor-imagick.php';
+	}
+
+	return array( 'Dominant_Color_Image_Editor_GD', 'Dominant_Color_Image_Editor_Imagick' );
+}
+
+/**
+ * Get dominant color of image
+ *
+ * @since n.e.x.t
+ *
+ * @param int $attachment_id the image id.
+ * @return string|WP_Error the dominant color of the image or WP_Error on error.
+ */
+function dominant_color_get_dominant_color( $attachment_id ) {
+	$file = get_attached_file( $attachment_id );
+	add_filter( 'wp_image_editors', 'dominant_color_set_image_editors' );
+	$editor = wp_get_image_editor( $file );
+	remove_filter( 'wp_image_editors', 'dominant_color_set_image_editors' );
+
+	if ( is_wp_error( $editor ) ) {
+		return $editor;
+	}
+	if ( ! method_exists( $editor, 'dominant_color_get_dominant_color' ) ) {
+		return new WP_Error( 'unable_to_find_method', __( 'Unable to find dominant_color_get_dominant_color method', 'performance-lab' ) );
+	}
+
+	$dominant_color = $editor->dominant_color_get_dominant_color();
+	if ( is_wp_error( $dominant_color ) ) {
+		return $dominant_color;
+	}
+
+	return $dominant_color;
+}
+
+/**
+ * Works out if color has transparency
+ *
+ * @since n.e.x.t
+ *
+ * @param int $id the attachment id.
+ * @return bool|WP_Error True if the color has transparency or WP_Error on error.
+ */
+function dominant_color_get_has_transparency( $id ) {
+	$file = get_attached_file( $id );
+	add_filter( 'wp_image_editors', 'dominant_color_set_image_editors' );
+	$editor = wp_get_image_editor( $file );
+	remove_filter( 'wp_image_editors', 'dominant_color_set_image_editors' );
+
+	if ( is_wp_error( $editor ) ) {
+		return $editor;
+	}
+	if ( ! method_exists( $editor, 'dominant_color_get_has_transparency' ) ) {
+		return new WP_Error( 'unable_to_find_method', __( 'Unable to find dominant_color_get_has_transparency method', 'performance-lab' ) );
+	}
```

Comment:
Most of this logic is duplicated from the above `dominant_color_get_dominant_color()` function and is furthermore a bit wasteful to run again, especially since the only place the two functions are called is right next to each other.

I think it would be more efficient to simplify these two to a single function, e.g. `dominant_color_get_dominant_color_and_transparency()`. Normally I wouldn't suggest combining functions like this, but they are only "wrappers" for the class methods anyway, and those can remain separate of each other.

I don't think there is any benefit of keeping these two functions separate though.

Reply:
Disagree here. This function can be called outside of this context and therefor needs this safeguard

Reply:
This code will all likely change on core merge / patch.

Reply:
> This function can be called outside of this context and therefor needs this safeguard

It can, but should it be? This function IMO should not be any encouraged API as it runs quite expensive logic. When plugin developers want to access dominant color or transparency values, they should use the metadata. We could even add wrapper functions that get the values from the meta data. But these two functions here compute the values on the fly, which is something external developers should not invoke.

Having two functions here is inefficient, and I don't see the benefit. Not every function is intended for third-party usage, even if other developers _can_ call it.